File: frontend/src/types/record.ts
export type RecordStatus = "Uploading" | "Processing" | "Ready" | "Failed";

export type CreateRecordRequest = {
  title: string;
  description: string;
  tags: string[];
};

export type CreateRecordResponse = {
  recordId: string;
  status: RecordStatus;
};

export type FinalizeRecordRequest = {
  recordId: string;
};

export type FinalizeRecordResponse = {
  recordId: string;
  status: "Processing" | "Ready" | "Failed";
  failureCode?: string | null;
  failureMessage?: string | null;
  retryable?: boolean | null;
};

export type RetryRecordResponse = {
  recordId: string;
  status: "Uploading";
  retryCount: number;
};

export type DeleteRecordResponse = {
  recordId: string;
  status: "Deleted" | "CleanupFailed";
  deletedArtifacts: number;
  failureCode?: string | null;
  failureMessage?: string | null;
  retryable?: boolean | null;
};

export type RecordDetailVideo = {
  url: string;
  duration: number | null;
  fps: number | null;
};

export type RecordDetailPose = {
  url: string;
  version: string;
};

export type RecordDetailMetricSummary = {
  metricId: string;
  unit: string | null;
  metricDefinitionVersion: string | null;
  activityType: string | null;
  side: string | null;
  min: number;
  max: number;
  average: number;
  rangeOfMotion: number;
};

export type RecordDetailMetrics = {
  seriesUrl: string | null;
  summary: RecordDetailMetricSummary[];
};

export type RecordDetail = {
  recordId: string;
  title: string;
  description: string;
  status: RecordStatus;
  video: RecordDetailVideo | null;
  pose: RecordDetailPose | null;
  metrics: RecordDetailMetrics | null;
  tags: string[];
  createdAt: string;
  updatedAt?: string | null;
  uploadingAt?: string | null;
  processingStartedAt?: string | null;
  readyAt?: string | null;
  failedAt?: string | null;
  failureStage?: string | null;
  failureCode?: string | null;
  failureMessage?: string | null;
  retryable?: boolean | null;
  retryCount?: number;
};

export type RecordListItem = {
  recordId: string;
  title: string;
  description: string;
  thumbnailUrl: string | null;
  duration: number | null;
  status: RecordStatus;
  tags: string[];
  createdAt: string;
  updatedAt?: string | null;
  failureCode?: string | null;
  failureMessage?: string | null;
  retryable?: boolean | null;
};

export type ListRecordsResponse = {
  items: RecordListItem[];
  total: number;
};

File: backend/app/schemas/record.py
from typing import Literal

from pydantic import BaseModel, Field


RecordStatus = Literal["Uploading", "Processing", "Ready", "Failed"]


class CreateRecordRequest(BaseModel):
    title: str = Field(min_length=1)
    description: str = ""
    tags: list[str] = Field(default_factory=list)
    duration: float | None = Field(default=None, gt=0)
    fps: float | None = Field(default=None, gt=0)


class CreateRecordResponse(BaseModel):
    recordId: str
    status: RecordStatus


class FinalizeRecordResponse(BaseModel):
    recordId: str
    status: Literal["Processing", "Ready", "Failed"]
    failureCode: str | None = None
    failureMessage: str | None = None
    retryable: bool | None = None


class RetryRecordResponse(BaseModel):
    recordId: str
    status: Literal["Uploading"]
    retryCount: int


class DeleteRecordResponse(BaseModel):
    recordId: str
    status: Literal["Deleted", "CleanupFailed"]
    deletedArtifacts: int
    failureCode: str | None = None
    failureMessage: str | None = None
    retryable: bool | None = None


class RecordDetailVideo(BaseModel):
    url: str
    duration: float | None = None
    fps: float | None = None


class RecordDetailPose(BaseModel):
    url: str
    version: str


class RecordDetailMetricSummary(BaseModel):
    metricId: str
    unit: str | None = None
    metricDefinitionVersion: str | None = None
    activityType: str | None = None
    side: str | None = None
    min: float
    max: float
    average: float
    rangeOfMotion: float


class RecordDetailMetrics(BaseModel):
    seriesUrl: str | None = None
    summary: list[RecordDetailMetricSummary] = Field(default_factory=list)


class RecordDetailResponse(BaseModel):
    recordId: str
    title: str
    description: str
    status: RecordStatus
    video: RecordDetailVideo | None = None
    pose: RecordDetailPose | None = None
    metrics: RecordDetailMetrics | None = None
    tags: list[str] = Field(default_factory=list)
    createdAt: str
    updatedAt: str | None = None
    uploadingAt: str | None = None
    processingStartedAt: str | None = None
    readyAt: str | None = None
    failedAt: str | None = None
    failureStage: str | None = None
    failureCode: str | None = None
    failureMessage: str | None = None
    retryable: bool | None = None
    retryCount: int = 0


class RecordListItem(BaseModel):
    recordId: str
    title: str
    description: str
    thumbnailUrl: str | None = None
    duration: float | None = None
    status: RecordStatus
    tags: list[str] = Field(default_factory=list)
    createdAt: str
    updatedAt: str | None = None
    failureCode: str | None = None
    failureMessage: str | None = None
    retryable: bool | None = None


class ListRecordsResponse(BaseModel):
    items: list[RecordListItem] = Field(default_factory=list)
    total: int
